Add makefile

meganwolf0 opened the pull request:
## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# Description

Currently the zarf packages are built and tested via Github actions. In an environment without access to those actions, the new `Makefile` and `build-package.sh` can be used to build and test those packages.

# How Has This Been Tested?

Had issues with total replication of Github actions, specifically when running the go tests. I think this largely had to do with the environment configuration, a couple different attempts were made:

- The Makefile was tested on a Amazon Linux instance with the following software packages installed: git, go, docker. Docker service was started prior to make commands and user added to docker user group (`usermod -aG docker ec2-user`). The EC2 instance had an IAM role allowed to pull ECR images from the requisite repo to build Zarf packages. Registry credentials were set as environmental variables. Size of /tmp to 8GB.
- Tested on EC2 instance defined by the cloud-init.yaml [here](https://github.com/naps-dev/pipeline-magic/blob/main/pl-runner/cloud-init.yaml), added a go install. Changed the ownership of the home directory to fedora (`sudo chown -R fedora:fedora .`). Unsure how the runner handles this, but to ssh in and run via command line this was required. Upped size of /tmp to 8GB.

Both instances failed when running the tests (target `run-tests`) so ultimately decided to remove this dependency from `all`. No concrete errors to point to, sometimes the instance just crashed, sometimes the tests timed out. These breakdowns seemed to happen around deployment of the dco-core package.

- [x] `make REF_NAME=v4.2.0-2 REF_TYPE=tag ZARF_PACKAGE=zarf-package-arkime-amd64.tar.zst IMAGE_TAG=v4.2.0-2 COMPONENT=arkime`
- [x] `make REF_NAME=v4.0.10-4 REF_TYPE=tag ZARF_PACKAGE=zarf-package-polarity-amd64.tar.zst COMPONENT=polarity POLARITY_LICENSE=$POLARITY_LICENSE`


**Test Configuration**:
* OS: Amazon Linux 2023 AMI 2023.0.20230614.0 x86_64 HVM kernel-6.1 -> Gave it 100 GB volume and needed to up the size of /tmp to 8GB handle Zarf logs spit out by the go tests
* IAM Role: naps-dev-ecr
* Go version: 1.19.9 

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have added a plain language explanation of the changes to the Unreleased section of the CHANGELOG.md file

bburky commented on Makefile:
```diff
@@ -0,0 +1,136 @@
+
+install-dependencies:
+# Zarf
+ifeq (,$(shell which zarf))
```
I've usually seen `command -v` recommended over `which` for checks on installed commands.

bburky commented on build-package.sh:
```diff
@@ -0,0 +1,27 @@
+rm -rf big-tmp
+mkdir big-tmp
+
+args=()
+if [ "$2" = "branch" ]; then
+    args+=(--set GIT_REF=refs/heads/$3)
+fi
+if [ "$2" = "tag" ]; then
+    args+=(--set GIT_REF=refs/tags/$3)
+fi
+if [ "$4" != "" ]; then
+    args+=(--set IMAGE_TAG="$4")
+fi
```
Maybe at least a one line comment on what the args are?

Might be better to test `$#` first and check if the argument count is 3 or 4. I think the current code has `$4` as sort of optional but would use an undefined variable if you check `[ "$4" != "" ]` without a 4th arg.

bburky reviewed (commented):
looks good overall

meganwolf0 commented on Makefile:
```diff
@@ -0,0 +1,136 @@
+	$(info "k3d already installed")
+endif
+
+install-zarf: 
+	curl -L -o zarf_v$(ZARF_VER)_Linux_amd64 https://github.com/defenseunicorns/zarf/releases/download/v$(ZARF_VER)/zarf_v$(ZARF_VER)_Linux_amd64
+	chmod +x ./zarf_v$(ZARF_VER)_Linux_amd64 && sudo mv ./zarf_v$(ZARF_VER)_Linux_amd64 /usr/local/bin/zarf
+	curl -L -o zarf-init-amd64-v$(ZARF_VER).tar.zst https://github.com/defenseunicorns/zarf/releases/download/v$(ZARF_VER)/zarf-init-amd64-v$(ZARF_VER).tar.zst
+	mkdir $(HOME)/.zarf-cache && mv ./zarf-init-amd64-v$(ZARF_VER).tar.zst $(HOME)/.zarf-cache/
+
+install-go: 
+	curl -L -o go.tar.gz https://dl.google.com/go/go$(GOLANG_VER_MIN).linux-amd64.tar.gz && \
+	tar -C /usr/local/ -xzf go.tar.gz && \
+	rm go.tar.gz
+
+install-aws:
+	curl "https://awscli.amazonaws.com/awscli-exe-linux-x86_64.zip" -o "awscliv2.zip" && \
+	unzip awscliv2.zip && \
+	sudo ./aws/install
```
Yeah I think that does make more sense - I struggled with the dependency management thing and defaulted to "just install it yourself" except in the case of zarf (and I guess k3d). I was having some issues with installing go on different OSs - for example when I tried to install on alpine it required some musl-dev lib, which seemed specific to go running on alpine? So I'm not sure how to handle some of those things that might crop up if this is running on different distros.

bburky replied:
Alpine is always very weird due to musl. The only easy way to avoid its is by using static binaries. Otherwise, yes you'd have to do some alpine specific things and actually use their OS packages.

docandrew replied:
Might be worth using the `zarf tools download-init` command in place of the curl for the init package as well but that can be modified later.

meganwolf0 commented:
Based on some feedback, I updated the Makefile as follows:
- Added the ZARF_CONFIG variable
- Renamed "ZARF_PACKAGE" to "ZARF_PACKAGE_NAME" to alleviate any weird Zarf variable things
- Added a target to build the packages in a Docker container, also added Dockerfile that builds the image
- Changed `which` to `command -v`
- Added arguments description and optional arg 4 checking in build-package.sh

Updated tests:
- `make build-all-docker` - requires .env
- `make build-all REF_NAME=v4.2.0-2 REF_TYPE=tag ZARF_PACKAGE_NAME=zarf-package-arkime-amd64.tar.zst IMAGE_TAG=v4.2.0-2 COMPONENT=arkime`

Minor clean-up/doc things
- Changed assignment type for some variables
- Moved "PHONYs" for readability
- Added help target
- Changed some target names
- Updated Makefile_README

Future work is probably more dependency clean up/installing and using local dependencies
If we like the docker thing, probably should push to ghcr, right now it just builds it locally

docandrew reviewed (approved):
Very thorough, I really appreciate the additional README - this is going to be a great starting point for decoupling build steps from our CI system.

jamesallen-atx reviewed (dismissed):
LGTM

benmountjoy111 reviewed (approved):
Looks great!

jamesallen-atx reviewed (approved):
👍
